src/rpc/stats.cpp

```diff
@@ -0,0 +1,203 @@
+#include <rpc/stats.h>
+
+#include <rpc/server.h>
+#include <rpc/util.h>
+
+UniValue RPCStats::toJSON() {
+    UniValue stats(UniValue::VOBJ),
+             latencyObj(UniValue::VOBJ),
+             payloadObj(UniValue::VOBJ),
+             historyArr(UniValue::VARR);
+
+    latencyObj.pushKV("min", latency.min);
+    latencyObj.pushKV("avg", latency.avg);
+    latencyObj.pushKV("max", latency.max);
+
+    payloadObj.pushKV("min", payload.min);
+    payloadObj.pushKV("avg", payload.avg);
+    payloadObj.pushKV("max", payload.max);
+
+    for (auto const &entry : history) {
+        UniValue historyObj(UniValue::VOBJ);
+        historyObj.pushKV("timestamp", entry.timestamp);
+        historyObj.pushKV("latency", entry.latency);
+        historyObj.pushKV("payload", entry.payload);
+        historyArr.push_back(historyObj);
+    }
+
+    stats.pushKV("name", name);
+    stats.pushKV("count", count);
+    stats.pushKV("lastUsedTime", lastUsedTime);
+    stats.pushKV("latency", latencyObj);
+    stats.pushKV("payload", payloadObj);
+    stats.pushKV("history", historyArr);
+    return stats;
+}
+
+RPCStats RPCStats::fromJSON(UniValue json) {
+    RPCStats stats;
+
+    stats.name = json["name"].get_str();
+    stats.lastUsedTime = json["lastUsedTime"].get_int64();
+    stats.count = json["count"].get_int64();
+    if (!json["latency"].isNull()) {
+        auto latencyObj  = json["latency"].get_obj();
+        stats.latency = {
+            latencyObj["min"].get_int64(),
+            latencyObj["avg"].get_int64(),
+            latencyObj["max"].get_int64()
+        };
+    }
+
+    if (!json["payload"].isNull()) {
+        auto payloadObj  = json["payload"].get_obj();
+        stats.payload = {
+            payloadObj["min"].get_int64(),
+            payloadObj["avg"].get_int64(),
+            payloadObj["max"].get_int64()
+        };
+    }
+
+    if (!json["history"].isNull()) {
+        auto historyArr = json["history"].get_array();
+        for (const auto &entry : historyArr.getValues()) {
+            auto historyObj = entry.get_obj();
+            StatHistoryEntry historyEntry {
+                historyObj["timestamp"].get_int64(),
+                historyObj["latency"].get_int64(),
+                historyObj["payload"].get_int64()
+            };
+            stats.history.push_back(historyEntry);
+        }
+    }
+    return stats;
+}
+
+void CRPCStats::add(const std::string& name, const int64_t latency, const int64_t payload)
+{
+    auto stats = CRPCStats::get(name);
+    if (stats) {
```

Comment:
No locks held during this point. It's mutated freely. And if a reference of the same stats is held from another method (see later), this can end up seeing partial writes resulting in races.

Reply:
Nope, there is no shared resource here. `CRPCStats::get` returns a copy in the stack which is not shared between threads.

Reply:
Reviewing again, the above statement is absolutely correct. They are all simple value types. Resolving this.

Reply:
Reviewing again, the above statement is absolutely correct. They are all simple value types. Resolving this.